packages/embed/index.tsx

```diff
@@ -0,0 +1,3 @@
+import ExcalidrawApp from "../../src/excalidraw-app/app";
```

Comment:
why rexport ? 
Let's add the entry file here itself. instead of re-exporting here.
This will create confusion since this is the npm package and the excalidraw app should import from here instead of maintaining the code outside this folder and it would be hard to isolate app and package code as well.

If you are doing this because of CRA then let's move packages inside the src folder instead.
cc @dwelle

Reply:
> If you are doing this because of CRA then let's move packages inside the src folder instead.

Not sure what CRA means but I did that because Vercel was complaining. I don't have the exact error message at hand anymore but it was not possible to reference something outside the _src_ directory.

> This will create confusion since this is the npm package and the excalidraw app should import from here instead of maintaining the code outside this folder and it would be hard to isolate app and package code as well.

Not sure about that since the code will be Webpack-ed the project (internal) structure won't really matter when you consume `@excalidraw/excalidraw` as a npm package.
From a developer/contributor perspective, any IDE will resolve the import so you will quickly see where the component is actually defined (i.e. in `src/excalidraw-app/app.tsx`).

It might be a personnel preference, but I like the fact that all packages are under "packages". My brain will have more work to do if the package definition of `@excalidraw/utils` was under "packages" but the package definition of `@excalidraw/excalidraw` was under "src".

Reply:
@guillaumebriday sorry I had edited your comment by mistake but now I have reverted it and posting my comments below. 

>Not sure what CRA means but I did that because Vercel was complaining. I don't have the exact error message at hand anymore but it was not possible to reference something outside the _src_ directory.

That's what I meant CRA(create react app) doesn't allow imports outside src directory :/ .

>Not sure about that since the code will be Webpack-ed the project (internal) structure won't really matter when you consume `@excalidraw/excalidraw` as a npm package.

It won't matter for consumers but as a good practice imo, it's better if we can name the directories according to the package name so, from the directory, it's clear which package is exported instead of checking package.json inside the directory.
 
>From a developer/contributor perspective, any IDE will resolve the import so you will quickly see where the component is actually defined (i.e. in `src/excalidraw-app/app.tsx`).


>It might be a personnel preference, but I like the fact that all packages are under "packages". My brain will have more work to do if the package definition of `@excalidraw/utils` was under "packages" but the package definition of `@excalidraw/excalidraw` was under "src".

imo it's not a personal preference but it will be unnecessary complicating the flow if we keep the code outside the packages folder and just re-export it coz of the fact that CRA doesn't allow imports outside src, the whole reason(better isolation) for moving inside packages sort of gets defeated and we end up adding an unnecessary index.js (for one entry) which just re-exports the entry file. 
And we are also trying to isolating a good chunk of code like firebase, local storage, collaboration which means that all these code will actually move to the excalidraw-app folder eventually and the rest will possibly move inside excalidraw-embed but with the current packages structure that won't be possible unless we keep the excalidraw-embed and excalidraw-app folders(which is redundant as per excalidraw-embed coz we can easily keep it inside packages) inside src and just import the entry file to export it from packages directory. 

And I agree we shouldn't keep some definitions outside `src` and some inside `src` so we should move the utils folder inside packages inside src in that case.

Reply:
> sorry I had edited your comment above by mistake woman_facepalming but now I have reverted it and posting my comments below.

No worries :smile: 

> It won't matter for consumers but as a good practice imo, it's better if we can name the directories according to the package name so, from the directory, it's clear which package is exported instead of checking package.json inside the directory.

I've renamed the directory in _packages_:

- `packages/excalidraw/package.json`
- `packages/utils/package.json`

> imo it's not a personal preference but it will be unnecessary complicating the flow if we keep the code outside the packages folder and just re-export it coz of the fact that CRA doesn't allow imports outside src

I thought we agreed on the _packages_ directory outside of _src_.
We decided to move `utils.ts` outside of _src_ to make it clear that this file should not be used/imported in the Excalidraw application.
We also decided to move "excalidraw-embed" outside of _src_ to keep the packages together. Arguably it's also a bit weird to have a package definition inside the _src_ directory.

I don't think there's a perfect solution here. Each solution has benefits and drawbacks but I think the current solution is the "least worst" and somehow future-proof.
What I mean by future-proof is that it will be possible to use/import the `@excalidraw/excalidraw` package from the React application (in the "src" folder).
In other words, the reexport is only is a transition step. In the future, `@excalidraw/excalidraw` will effectively contains code.

> And we are also trying to isolating a good chunk of code like firebase, local storage, collaboration which means that all these code will actually move to the excalidraw-app folder eventually and the rest will possibly move inside excalidraw-embed but with the current packages structure that won't be possible unless we keep the excalidraw-embed and excalidraw-app folders(which is redundant as per excalidraw-embed coz we can easily keep it inside packages) inside src and just import the entry file to export it from packages directory.

I think the React application should stay in "src" (Firebase, local storage, collaboration...) and the rest should move to the `@excalidraw/excalidraw`.
But I also don't want to do that in this pull request. It feels like we are trying to do too much in this pull request. This pull request is *not* about "excalidraw-embed".

> That's what I meant CRA(create react app) doesn't allow imports outside src directory :/ .

It seems a bit arbitrary but as mentioned above, I think it won't be an issue once we declare `@excalidraw/excalidraw` as a dependency of the React application.

Anyway, we should focus on `@excalidraw/utils` not on `@excalidraw/excalidraw`.

Reply:
>I thought we agreed on the packages directory outside of src.

Yes, we agreed but I was under the impression that the whole excalidraw-embed folder will be moving inside the packages and didn't realize CRA would not allow. But since we are bound by CRA I think we should keep it inside `src` and later if plan to import it as a package then we can move it outside src

>We decided to move utils.ts outside of src to make it clear that this file should not be used/imported in the Excalidraw application.

Yes but as said above I didn't think about the CRA that time and as long as we are not consuming it as packages I think it will be tough to move it outside src. 
Also, I think we need not worry too much unless there is some major concern if kept inside `src`

>We also decided to move "excalidraw-embed" outside of src to keep the packages together. Arguably it's also a bit weird to have a package definition inside the src directory.

yeah but as mentioned above as long as we are bound by CRA or we are not consuming it as packages we can't do much about it :/ 


>I don't think there's a perfect solution here. Each solution has benefits and drawbacks but I think the current solution is the "least worst" and somehow future-proof.
What I mean by future-proof is that it will be possible to use/import the @excalidraw/excalidraw package from the React application (in the "src" folder).
In other words, the reexport is only is a transition step. In the future, @excalidraw/excalidraw will effectively contains code.

Yes, when we plan to consume it as a package that time we should definitely move it outside src but till then we should keep inside `src`. It's sad but sometimes CRA does make things tough :(  

>I think the React application should stay in "src" (Firebase, local storage, collaboration...) and the rest should move to the @excalidraw/excalidraw.

Yes, that's what I mentioned above :). 

>But I also don't want to do that in this pull request. It feels like we are trying to do too much in this pull request. This pull request is not about "excalidraw-embed".

I didn't mean to make the changes in this PR for this, we are already working on it so soon the code for excalidraw-embed and excalidraw-app will be isolated :).
But as I mentioned above imo it would be better if code stays inside `src/packages/excalidraw` instead of `src/excalidraw-embed` and then importing the entry point inside `packages/excalidraw` just to re-export it. 
This way the code also stays inside respective packages without any redundancy and later if we want to consume it as package we can move the entire packages outside src as it is as well.

>It seems a bit arbitrary but as mentioned above, I think it won't be an issue once we declare @excalidraw/excalidraw as a dependency of the React application.

Yes as mentioned once we start consuming it as package then only we should move it outside src.

>Anyway, we should focus on @excalidraw/utils, not on @excalidraw/excalidraw

Yep, but moving packages outside src was affecting `excaldraw-embed` hence had to point that out.

Reply:
> But as I mentioned above imo it would be better if code stays inside src/packages/excalidraw instead of src/excalidraw-embed and then importing the entry point inside packages/excalidraw just to re-export it.

Ok, let me try to summarize, so we would have the following directory structure:

```
packages
└── utils
src
├── actions
├── components
├── css
├── data
├── element
├── excalidraw-app
├── locales
├── packages
│   └── excalidraw
├── renderer
├── scene
└── tests
    ├── fixtures
    ├── helpers
    ├── queries
    └── __snapshots__
```

`src/packages/excalidraw/index.tsx` will contain the Excalidraw component definition:

```js
// ...

const Excalidraw = (props: ExcalidrawProps) => {
  //...
}

// ...
```

`src/excalidraw-app/index.tsx` will import the Excalidraw component:

```js
// ...

import Excalidraw from "../packages/excalidraw/index";

// ...
```

Is that what you have in mind?

@dwelle @lipis is everyone OK with that? I prefer asking before moving the files again :sweat_smile:

Reply:
Please note that with the above directory structure, the package `@excalidraw/excalidraw` declared in `src/packages/excalidraw` won't be able to export functions from the `packages/utils.ts` file because CRA won't let us import something from outside the _src_ directory.

Not sure if it was a goal but I want to state every potential issues.

Reply:
@Mogztter as explained in the above discussion I  meant let's have `packages` inside `src` so `utils` will also go inside `packages` inside `src`. Having two packages is even more confusing.

Reply:
haven't been following the discussion... why do we want to have `packages`? That's not a good idea at least from intellisense perspective. Lemme read up on the thread here.

Reply:
I see, due to CRA. Yea, so we have 2 options. Move `packages` to `src` as @ad1992 suggested, or migrate away from CRA :).

Reply:
We should be able to solve the intellisense thing by `exclude` in `tsconfig`.

Reply:
>  as explained in the above discussion I meant let's have packages inside src so utils will also go inside packages inside src.

Ok, so back to square one :wink: 

```
src
├── actions
├── components
├── css
├── data
├── element
├── excalidraw-app
├── locales
├── packages
│   ├── excalidraw
│   └── utils
├── renderer
├── scene
└── tests
    ├── fixtures
    ├── helpers
    ├── queries
    └── __snapshots__
```

The file `utils.ts` will be declared in `src/packages`.

@dwelle @lipis @ad1992 everyone is OK with the above structure?

> We should be able to solve the intellisense thing by exclude in tsconfig.

I'm not familiar with IntelliSense, do you have an example?

Reply:
Sounds good to me.

> I'm not familiar with IntelliSense, do you have an example?

Np, I'll make the change myself after you settle the code.

Reply:
Perfect, I will update the code accordingly.
Thanks for your review!